internal/database/database.go

```diff
@@ -1188,11 +1192,21 @@ func (d *Database) UpsertSource(ctx context.Context, principal string, source *c
 			}
 		}
 
+		// Upsert requirements
+
+		var sources []string
 		for _, r := range source.Requirements {
 			if r.Source != nil {
 				if err := d.upsert(ctx, tx, "sources_requirements", []string{"source_name", "requirement_name", "gitcommit"}, []string{"source_name", "requirement_name"}, source.Name, r.Source, r.Git.Commit); err != nil {
 					return err
 				}
+				sources = append(sources, *r.Source)
+			}
+		}
+
+		if source.Requirements != nil {
+			if err := d.deleteNotIn(ctx, tx, "sources_requirements", "source_name", source.Name, "requirement_name", sources); err != nil {
+				return err
 			}
```

Comment:
This looks decent. What I started wondering is if a plain "delete from sources_requirements where source_name = ...." followed by an insert of the current ones wouldn't be simpler. Well, it most likely _is_ simpler, but I guess since it's two calls either way.... potato, potato.